### DISCUSSION TITLE
Intelligent switching in `lsp/quick/visual` scene with only one shortcut key

### DISCUSSION THREAD

**Discussion by xlboy** [johmsalas/text-case.nvim#153] [Show and tell]
### Example

https://github.com/johmsalas/text-case.nvim/assets/63690944/44c06f3c-81bb-4c79-8053-2ea2bd9e4882

### Code

```lua
local function smart_open_textcase()
  local telescope_extensions = require("telescope").extensions
  local telescope_opts = {
    prompt_title = "Text Case - %s",
    layout_strategy = "cursor",
    layout_config = { cursor = { width = 40, height = 16 } },
  }

  local is_visual = vim.api.nvim_get_mode().mode == "v"
  if is_visual then
    local tc_state = require("textcase.plugin.plugin").state
    local tc_utils = require("textcase.shared.utils")

    tc_state.telescope_previous_visual_region =
      tc_utils.get_visual_region(0, true, nil, tc_utils.get_mode_at_operator("v"))
    telescope_opts.prompt_title = string.format(telescope_opts.prompt_title, "Visual")
    telescope_extensions.textcase.visual_mode(telescope_opts)
    return
  end

  vim.lsp.buf_request(0, "textDocument/prepareRename", vim.lsp.util.make_position_params(), function(err, result)
    local can_rename = result ~= nil
    local fn_name = can_rename and "normal_mode_lsp_change" or "normal_mode_quick_change"
    telescope_opts.prompt_title = string.format(telescope_opts.prompt_title, can_rename and "LSP" or "Quick")
    telescope_extensions.textcase[fn_name](telescope_opts)
  end)
end

-- in lazy.nvim
return {
  {
    "johmsalas/text-case.nvim",
    keys = {
      { "gai", smart_open_textcase, desc = "[Text Case] Smart Open", mode = { "n", "v" } },
    },
    config = function()
      require("textcase").setup({ default_keymappings_enabled = false })
      require("telescope").load_extension("textcase")
    end,
  },
}

--------------------------------------------------------------

-- or manual setting

-- 1. config:
-- require("textcase").setup({ default_keymappings_enabled = false })
-- require("telescope").load_extension("textcase")

-- 2. keymap:
-- vim.keymap.set({ "n", "v" }, "gai", smart_open_textcase)

```

**Comment by johmsalas** [comment]
Hi!. Timing for this idea is perfect, we are currently working on `v2 - Ergonomics` and I've been wondering how to simplify the Telescope menu. I like your proposal on detecting the context. Opening threads for discussion

**Comment by johmsalas** [comment]
In the case of a variable, the user should be able to freely select if they want to apply LSP or Quick rename. I see 2 solutions:
- Keep LSP and Quick change as separated bindings
- Or integrate both into a single small list, and provide two different keybindings for selecting the item. ie, <enter> will apply the selected item using Word Under Cursor and "!" will apply LSP renaming

**Reply by xlboy** [reply]
I tend towards having both:

1. Keep the key bindings for LSP/Quick, leaving freedom to the users.

2. When on a variable:
    - A list can be popped up for users to select from.
    - The default can be made to use LSP/Quick based on user configuration. (If I am performing a rename operation on a variable, I will only use LSP)

    ```lua
    local config = {
      -- ...
      smart_open = {
        enabled = true,
        mode_at_variable = "lsp", --- @type "lsp" | "quick" | "popup-selection"
        keymap = "gai"
      }
      -- ...
    }
    ```

**Reply by johmsalas** [reply]
Yeah, agree. Why not both

**Comment by johmsalas** [comment]
Do you plan to create a PR? replacing normal_mode_lsp_change is a good start, at the end LSP is default and there would be graceful degradation to quick renaming when LSP is not available

**Reply by xlboy** [reply]
I'm sorry, I've been busy with other things recently and haven't had the time to contribute.

Regarding "replacement", I still think it needs further consideration. If it were up to me, I would prefer to keep all the sub-APIs but provide an additional "Smart Open" function.

- It caters to the habits of previous users.
  
- It leaves the freedom to the users.
> In fact, I'm a person who values freedom a lot.

---

Although, with this "Smart Open" function, many users may no longer need the `normal_mode_lsp_change`, but there might still be a few unique users who will use it? (To be determined...)

**Reply by johmsalas** [reply]
Some warnings so that you can take an informed decision. 
* There are ways to find out if the word under the cursor is a definition, using LSP hover or treesitter
* Visual mode needs 2 checks. "v" and "\22"  (visual block)
* it might require some integration tests. I can provide those to make your experience as a contributor smoother.

No worries, however you feel more comfortable. in If you give your permission I can add it for you otherwise we can wait. Whatever you consider is fine. Btw, Appreciate your contribution I hadn't thought on this specific idea.

**Reply by xlboy** [reply]
Haha, thank you very much for your support towards me and your outstanding contribution to this plugin. If you haven't added this feature in the future and I have some free time to contribute, I would be very willing to lend my strength to this cause. :)
